device/tt_soc_descriptor.cpp

```diff
@@ -336,18 +348,33 @@ void tt_SocDescriptor::get_cores_and_grid_size_from_coordinate_manager() {
     }
 }
 
-std::vector<tt::umd::CoreCoord> tt_SocDescriptor::get_cores(const CoreType core_type) const {
-    if (cores_map.find(core_type) == cores_map.end()) {
-        return {};
+void tt_SocDescriptor::translate_coordinates(
+    std::map<std::pair<CoreType, CoordSystem>, std::vector<CoreCoord>> &core_coord_map,
+    const CoreType core_type,
+    const CoordSystem coord_system) const {
+    std::vector<CoreCoord> cores = core_coord_map.at({core_type, CoordSystem::PHYSICAL});
+    std::vector<CoreCoord> translated_cores;
+    for (const auto &core : cores) {
+        translated_cores.push_back(translate_coord_to(core, coord_system));
+    }
+    core_coord_map.insert({{core_type, coord_system}, translated_cores});
+}
+
+std::vector<tt::umd::CoreCoord> tt_SocDescriptor::get_cores(
+    const CoreType core_type, const CoordSystem coord_system) const {
+    if (cores_map.find({core_type, coord_system}) == cores_map.end()) {
+        translate_coordinates(cores_map, core_type, coord_system);
     }
-    return cores_map.at(core_type);
+    return cores_map.at({core_type, coord_system});
 }
 
-std::vector<tt::umd::CoreCoord> tt_SocDescriptor::get_harvested_cores(const CoreType core_type) const {
-    if (harvested_cores_map.find(core_type) == harvested_cores_map.end()) {
-        return {};
+std::vector<tt::umd::CoreCoord> tt_SocDescriptor::get_harvested_cores(
+    const CoreType core_type, const CoordSystem coord_system) const {
+    log_assert(coord_system != CoordSystem::LOGICAL, "Harvested cores are not supported for logical coordinates");
+    if (harvested_cores_map.find({core_type, coord_system}) == harvested_cores_map.end()) {
+        translate_coordinates(harvested_cores_map, core_type, coord_system);
```

Comment:
Note that this call is not thread safe... Lazy cache needs to be locked (both read and write) for thread safety....

Reply:
Lazy buffering is removed, now everything is read only